pkg/cryptoauthlib/psa_se_driver/psa_atca_se_driver.c

```diff
@@ -0,0 +1,561 @@
+     (type == PSA_KEY_TYPE_HMAC && size == 32))
+
+/**
+ * @brief   Convert ATCA status values to PSA errors
+ *
+ * @param   error
+ *
+ * @return  @ref psa_status_t
+ */
+static psa_status_t atca_to_psa_error(ATCA_STATUS error)
+{
+    switch (error) {
+    case ATCA_NOT_LOCKED:
+    case ATCA_EXECUTION_ERROR:
+    case ATCA_FUNC_FAIL:
+        return PSA_ERROR_BAD_STATE;
+    case ATCA_WAKE_FAILED:
+    case ATCA_RX_FAIL:
+    case ATCA_RX_NO_RESPONSE:
+    case ATCA_TX_TIMEOUT:
+    case ATCA_RX_TIMEOUT:
+    case ATCA_TOO_MANY_COMM_RETRIES:
+    case ATCA_COMM_FAIL:
+    case ATCA_TIMEOUT:
+    case ATCA_TX_FAIL:
+        return PSA_ERROR_COMMUNICATION_FAILURE;
+    case ATCA_RX_CRC_ERROR:
+    case ATCA_STATUS_CRC:
+        return PSA_ERROR_DATA_CORRUPT;
+    case ATCA_SMALL_BUFFER:
+        return PSA_ERROR_BUFFER_TOO_SMALL;
+    case ATCA_BAD_OPCODE:
+    case ATCA_BAD_PARAM:
+    case ATCA_INVALID_SIZE:
+    case ATCA_INVALID_ID:
+        return PSA_ERROR_INVALID_ARGUMENT;
+    case ATCA_UNIMPLEMENTED:
+        return PSA_ERROR_NOT_SUPPORTED;
+    default:
+        return PSA_ERROR_GENERIC_ERROR;
+    }
+}
+
+/* Secure Element Cipher Functions */
+/**
+ * @brief   Set up a driver specific AES CBC mode operation
+ *
+ * @param   ctx
+ * @param   dev
+ * @param   key_slot
+ */
+static void atca_cbc_setup(atca_aes_cbc_ctx_t *ctx,
+                           ATCADevice dev,
+                           psa_key_slot_number_t key_slot)
+{
+    ctx->device = dev;
+    ctx->key_id = key_slot;
+    ctx->key_block = 0;
+}
+
+psa_status_t atca_cipher_setup(psa_drv_se_context_t *drv_context,
+                               void *op_context,
+                               psa_key_slot_number_t key_slot,
+                               psa_algorithm_t algorithm,
+                               psa_encrypt_or_decrypt_t direction)
+{
+    ATCADevice dev = (ATCADevice)drv_context->transient_data;
+    psa_se_cipher_context_t *ctx = (psa_se_cipher_context_t *)op_context;
+
+    /* Only device type ATECC608 supports AES operations */
+    if (dev->mIface.mIfaceCFG->devtype != ATECC608) {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    /* This implementation is for demonstration and currently only supports AES ECB encryption */
+    if (!ALG_IS_SUPPORTED(algorithm)) {
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    switch (algorithm) {
+    case PSA_ALG_CBC_NO_PADDING:
+        atca_cbc_setup(&ctx->drv_ctx.atca_aes_cbc, dev, key_slot);
+        break;
+    default:
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+    ctx->direction = direction;
+
+    return PSA_SUCCESS;
+}
+
+psa_status_t atca_cipher_set_iv(void *op_context,
+                                const uint8_t *p_iv,
+                                size_t iv_length)
+{
+
+    atca_aes_cbc_ctx_t *ctx = &((psa_se_cipher_context_t *)op_context)->drv_ctx.atca_aes_cbc;
+
+    if (iv_length != ATCA_MAX_IV_LEN) {
+        return PSA_ERROR_INVALID_ARGUMENT;
```

Comment:
This might profit from a `DEBUG("iv_length is not equal ATCA_MAX_IV_LEN)"` print. Otherwise the `PSA_ERROR_INVALID_ARGUMENT` leaves you wondering, which of the three arguments is invalid? Additionally, why do they have to be the same length? Followed up by: Why is the argument needed, if it is always the same value?

Reply:
It is only the same value for this device. Theroretically we could have a second Secure Element that supports ciphers with a different IV length. 
So it is possible that a user accidentally wants to perform a differen cipher operation, but passes the wrong location or something like that, so we need to check the size at runtime.
But I'll add the debug print.